### PR TITLE
Implement Service SAS generation

### LINKED ISSUE #66: Add support for generation of SAS tokens/signed URLs
We have some code to generate service SAS that we're willing to contribute back to this repo. What would be the best place to put a method to generate service SAS tokens?

Also, @jcorioland published [a gist](https://gist.github.com/jcorioland/731d7ab2aac9b505120a1a80a0188692) with code to generate Account SAS tokens that would be helpful to have as part of the SDK too.

### PR DESCRIPTION
**Pull request by dluces**
Fixes #66

WARNING: The tests added here are WRONG. I have no easy way of generating actual service SAS tokens right now, so the expected values in these tests were generated by using the tested method. This is NOT complete. If someone could point me to a service to generate actual tokens, it would be great.

### CODE REVIEW

**Comment by katmsft** [comment]
@dluces Please resolve the conflicts first. Also, if you could reference the newly added `tests/functional/Common/AccountSASFunctionalTest.php` that would be awesome.

File: tests/unit/Common/Internal/Authentication/SharedAccessSignatureHelperTest.php [hunk 1]
```diff
@@ -233,4 +233,377 @@ public function testGenerateAccountSharedAccessSignatureToken()
             $this->assertEquals($testCase[8], urlencode($actualSignature));
         }
     }
+
+    public function testGenerateServiceSharedAccessSignatureToken()
+    {
+        // Setup
+        $accountName = "phptests";
+        $accountKey = "WaZvixrkMok53QDmj8Tc99+BV6vO9cCOJNsdm+wD9QVEScwl8c1eYPcQ182ndNFqxX1+SEKs18SmOxh8OpzIUg==";
```
Comment (katmsft) [inline tests/unit/Common/Internal/Authentication/SharedAccessSignatureHelperTest.php]: Assume that this is a fake key. There is already one [here](https://github.com/Azure/azure-storage-php/blob/master/tests/framework/TestResources.php#L54), Please use this one.

File: tests/unit/Common/Internal/Authentication/SharedAccessSignatureHelperTest.php [hunk 2]
```diff
@@ -233,4 +233,377 @@ public function testGenerateAccountSharedAccessSignatureToken()
...
+                $testCase->getContentLanguage(),
+                $testCase->getContentType(),
+                $testCase->getStartingPartitionKey(),
+                $testCase->getStartingRowKey(),
+                $testCase->getEndingPartitionKey(),
+                $testCase->getEndingRowKey()
+            );
+
+            // assert
+            $this->assertEquals($testCase->getExpectedSignature(), urlencode($actualSignature));
```
Comment (katmsft) [inline tests/unit/Common/Internal/Authentication/SharedAccessSignatureHelperTest.php]: Notice that this needs to be changed once the mock key has been changed.

File: tests/unit/Common/Internal/Authentication/SharedAccessSignatureHelperTest.php [hunk 3]
```diff
@@ -233,4 +233,377 @@ public function testGenerateAccountSharedAccessSignatureToken()
...
+            "https", // signedProtocol
+            "", // signedIP
+            "k1BKI65TdXs7rdAJiqDSJ6wYHjfJD0CJgplvOyqBK7Y%3D" // expectedSignature
+        );
+
+        return $testCases;
+    }
+}
+
+class GenerateServiceSASTestCase extends GenerateAccountSASTestCase
```
Comment (katmsft) [inline tests/unit/Common/Internal/Authentication/SharedAccessSignatureHelperTest.php]: Advice not to have two class in one file, and also, it does not make much sense for this class to inherit from `GenerateAccountSASTestCase`. Also please reference the updated test case for account SAS [here](https://github.com/Azure/azure-storage-php/blob/master/tests/unit/Common/Internal/Authentication/SharedAccessSignatureHelperTest.php#L207) for a better practise.

**Comment by katmsft** [comment]
@hasonmsft, Please take a look.

File: src/Common/SharedAccessSignatureHelper.php [hunk 4]
```diff
@@ -64,6 +64,156 @@ public function __construct($accountName, $accountKey)
...
+     * @param  string $signedProtocol     Signed protocol.
+     * @param  string $cacheControl       Cache-Control header (rscc).
+     * @param  string $contentDisposition Content-Disposition header (rscd).
+     * @param  string $contentEncoding    Content-Encoding header (rsce).
+     * @param  string $contentLanguage    Content-Language header (rscl).
+     * @param  string $contentType        Content-Type header (rsct).
+     *
+     * @return string                     Service Shared Access Signature.
+     */
+    public function generateServiceSharedAccessSignatureToken(
```
Comment (hasonmsft) [inline src/Common/SharedAccessSignatureHelper.php]: Suggest to split this function to 4 functions and put them into the service proxies classes, e.g. BlobRestProxy. 

Several concerns:

1. Content-* headers are only for blob & file services only. 
2. For each services, the allowed permission are different. 
3. startpk, startrk, endpk, endrk only applicable for table service. 

Combining all the parameters in the same function will make users confused. :)

File: src/Common/SharedAccessSignatureHelper.php [hunk 5]
```diff
@@ -64,6 +64,156 @@ public function __construct($accountName, $accountKey)
     }
 
     /**
+     * Generates a shared access signature at the service level.
+     *
+     * This only supports version 2015-04-05 and later.
+     *
+     * @link https://msdn.microsoft.com/library/dn140255.aspx Constructing a service SAS.
```
Comment (hasonmsft) [inline src/Common/SharedAccessSignatureHelper.php]: @link looks good to me if it's an external URI format.

File: src/Common/SharedAccessSignatureHelper.php [hunk 6]
```diff
@@ -64,6 +64,156 @@ public function __construct($accountName, $accountKey)
...
+
+        // check that signed IP is valid
+        Validate::isString($signedIP, 'signedIP');
+
+        // validate and sanitize signed protocol
+        $signedProtocol = $this->validateAndSanitizeSignedProtocol($signedProtocol);
+
+        // check that signed identifier is valid
+        Validate::isString($signedIdentifier, 'signedIdentifier');
+        Validate::isTrue(
```
Comment (hasonmsft) [inline src/Common/SharedAccessSignatureHelper.php]: If the signedIdentifier is provided, some other parameter is not allowed - start, expiry, permission.

**Comment by dluces** [comment]
Hello. Yes, this is still a work in progress. I had to rebase off of dev and just realized that the auto merge changed some stuff. Thanks for the comments, I'll be updating this

File: tests/unit/Common/Internal/Authentication/SharedAccessSignatureHelperTest.php (hunk 1 above)
Comment (dluces) [inline tests/unit/Common/Internal/Authentication/SharedAccessSignatureHelperTest.php]: It is, I was using the same info that the Account SAS PR was using. It got changed later. I'll have to update all these references. Thanks

**Comment by katmsft** [comment]
Hi @dluces , Any updates recently? This is a feature request for many of our users so if you could not help us merge this PR before 0.16.0, which is aim to be available early in June, we will close this PR and implement this feature ourselves. But again, thanks for your contribution.

**Comment by dluces** [comment]
I'm really sorry, @katmsft. Things over here have been crazy busy and we had to prioritize contributions that were a blocker for us. This PR needs a lot of refactoring right now so I will close it. I don't think I'll be able to finish it for early June. Sorry!

**Pull request closed without merging**